Migrated bring your own Kafka connector (#644)
aiven/devportal: docs/products/kafka/kafka-connect/howto/bring-your-own-kafka-connect-cluster.rst

* ``PG_USERNAME``: The PostgreSQL service username
* ``PG_PASSWORD``: The PostgreSQL service password
* ``PG_DATABASE_NAME``: The PostgreSQL service database name

.. Note::

    If you're using Aiven for PostgreSQL and Aiven for Apache Kafka the above details are available in the `Aiven console <https://console.aiven.io/>`_ service Overview tab or via the dedicated ``avn service get`` command with the :ref:`Aiven CLI <avn_service_get>`.

Attach your own Apache Kafka Connect cluster to Aiven for Apache Kafka®
-----------------------------------------------------------------------

The following example demonstrates how to setup a local Apache Kafka Connect cluster with a working JDBC sink connector and attach it to an Aiven for Apache Kafka service.

.. _setup_trustore_keystore_bring_your_own_connect:

Setup the truststore and keystore
''''''''''''''''''''''''''''''''''

Create a :doc:`Java keystore and truststore <../../howto/keystore-truststore>` for the Aiven for Apache Kafka service.
For the following example we assume:

* The keystore is available at ``KEYSTORE_PATH/client.keystore.p12``
* The truststore is available at ``TRUSTSTORE_PATH/client.truststore.jks``

> * The same secret (password) is used for both the keystore and the truststore, and is shown here as ``KEY_TRUST_SECRET``

Reviewer: Thanks @TibsAtWork.
From a security standpoint is not ideal to have the same passwords, but I wouldn't loose too much time into the details of keystores since it's not the main topic of the article which is very long already. WDYT?

> * The same secret (password) is used for both the keystore and the truststore, and is shown here as ``KEY_TRUST_SECRET``

Reviewer 2: I agree it's a bit late in the process to be fiddling with changes of that scope.
Maybe add a "For simplicity," in front of "the same ..." (which feels naughty, but at least hints we know we're not being good).

> * The same secret (password) is used for both the keystore and the truststore, and is shown here as ``KEY_TRUST_SECRET``

Reviewer: For simplicity works!

1. Enable the **Schema Registry (Karapace)** and **Apache Kafka REST API (Karapace)**

2. In the **Topic** tab, create a new topic called ``jdbc_sink``, the topic will be used by the Apache Kafka Connect connector

Download the required binaries
''''''''''''''''''''''''''''''

The following binaries are needed to setup a Apache Kafka Connect cluster locally:

* `Apache Kafka <https://kafka.apache.org/quickstart>`_
* `Aiven Kafka connect JDBC connector <https://github.com/aiven/jdbc-connector-for-apache-kafka/releases>`_
* If you are going to use Avro as the data format, `Avro Value Converter <https://www.confluent.io/hub/confluentinc/kafka-connect-avro-converter>`_. The examples below show how to do this.

Setup the local Apache Kafka Connect cluster
''''''''''''''''''''''''''''''''''''''''''''

The following process defines the setup required to create a local Apache Kafka Connect cluster. The example shows the steps needed with the Apache Kafka ``3.1.0``, Avro converter ``7.1.0`` and JDBC connector ``6.7.0`` versions:

1. Extract the Apache Kafka binaries

>    .. code-block:: shell

Reviewer: Perfect thanks!

3. Check the status of the JDBC sink connector instance, ``jq`` is used to beautify the output

.. code-block:: shell

   curl localhost:8083/connectors/jdbc-sink-pg/status | jq

The result should be similar to the following

.. code-block:: json

   {
      "name": "jdbc-sink-pg",
      "connector": {
         "state": "RUNNING",
         "worker_id": "10.128.0.12:8083"
      },
      "tasks": [
         {
            "id": 0,
            "state": "RUNNING",
            "worker_id": "10.128.0.12:8083"
         }
      ],
      "type": "sink"
   }

Verify the JDBC connector using Karapace REST APIs
''''''''''''''''''''''''''''''''''''''''''''''''''

To verify that the connector is working, you can write messages to the ``jdbc_sink`` topic in Avro format using `Karapace REST APIs <https://github.com/aiven/karapace>`_ by following the steps below:

1. Create a new **Avro schema** using the ``/subjects/`` endpoint, after changing the placeholders for ``REST_API_USER``, ``REST_API_PASSWORD``, ``APACHE_KAFKA_HOST``, ``REST_API_PORT``

.. code-block:: shell

   curl -X POST -H "Content-Type: application/vnd.schemaregistry.v1+json" \
      --data '''
         {"schema":

>             "{\"type\": \"record\",\"name\": \"jdbcsinkexample\",\"namespace\": \"example\",\"doc\": \"example\",\"fields\": [{ \"type\": \"string\", \"name\": \"name\", \"doc\": \"person name\", \"namespace\": \"example\", \"default\": \"mario\"},{ \"type\": \"int\", \"name\": \"age\", \"doc\": \"persons age\", \"namespace\": \"example\", \"default\": 5}]}"

Reviewer 2: In the final text, this loses a lot of information off the right hand side. I'm assuming the reader would like to be able to read it all (I do note that *copying* it gives all the text correctly).

Reviewer: The main problem is that you can't easily split the strings into multiple lines (or at least I couldn't)... do you have a suggestion for it?

Reviewer 2: Hmm. I'll spend (no more than) 5 minutes on it after lunch - if it's not very easy I agree we shouldn't worry.

Reviewer 2: Hmm. Given the text can be copied, and given there's a scrollbar, and given the only simple thing I can think of is to put the text into a JSON file and use `--data @jsonfile`, which would mean changing the text and then testing again, I think we just leave it as is.
The intent of the command is clear, and explained clearly, and the command as shown doesn't take too many lines, so I think this is OK.

         }''' \
      https://REST_API_USER:REST_API_PASSWORD@APACHE_KAFKA_HOST:REST_API_PORT/subjects/jdbcsinkexample/versions/

The above call creates a new schema called ``jdbcsinkexample`` with a schema containing two fields (``name`` and ``age``).

2. Create a new **message** in the ``jdbc_sink`` topic using the ``jdbcsinkexample`` schema, after changing the placeholders for ``REST_API_USER``, ``REST_API_PASSWORD``, ``APACHE_KAFKA_HOST``, ``REST_API_PORT``

.. code-block:: shell

   curl -H "Content-Type: application/vnd.kafka.avro.v2+json" -X POST \
      -d '''
         {"value_schema": 

>             "{\"namespace\": \"test\", \"type\": \"record\", \"name\": \"example\", \"fields\": [{\"name\": \"name\", \"type\": \"string\"},{\"name\": \"age\", \"type\": \"int\"}]}", 

Reviewer 2: Same comment as above about not seeing a lot of the text

Reviewer 2: As previous comment.

> Bring your own Kafka connect cluster

Reviewer 3: ```suggestion
Bring your own Kafka Connect cluster
```

====================================

> Aiven provides Kafka connect cluster as a managed service in combination
> with Aiven for Apache Kafka managed service. However, there are

Reviewer 3: ```suggestion
Aiven provides a Kafka connect cluster as a managed service in combination
with the Aiven for Apache Kafka managed service. However, there are
```

circumstances where you may want to roll your own Connect cluster. This
help article shows the steps necessary to integrate your own connect
cluster with Aiven for Apache Kafka (and Schema registry). In this
example we create a JDBC sink connector to PostgreSQL database.

**Note:** As of version 3.0, Aiven for Apache Kafka no longer supports
Confluent Schema Registry. For more information, see `this article that
describes the replacement,

> Karapace <https://help.aiven.io/en/articles/5651983>`__ .

Reviewer 3: ```suggestion
Karapace <https://help.aiven.io/en/articles/5651983>`__.
```

As a prerequisite, the following information should be collected.

.. _following-aiven-kafka-and-postgresql-services-details-are-required-from-aiven-console-for-the-respective-service:

Following Aiven for Apache Kafka and PostgreSQL services' details are required from Aiven console for the respective service:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

| *KAFKA_HOST*
| *KAFKA_PORT*
| *SCHEMA_REGISTRY_PORT*
| *SCHEMA_REGISTRY_PW*
| *KAFKA_CONNECT_SERVICE_URI*
| *PG_SERVICE_URI*
| *PG_HOST*
| *PG_PORT*
| *PG_PW*

> Setup truststore and keystore:

Reviewer 3: ```suggestion
Setup the truststore and keystore:
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#. Download the Aiven project CA and client keys from the console

#. Generate *truststore* and *keystore*

::

   openssl pkcs12 -export -inkey service.key -in service.cert -out client.keystore.p12 -name service_key

   keytool -import -file ca.pem -alias CA -keystore client.truststore.jks 

Setup the Kafka service:
~~~~~~~~~~~~~~~~~~~~~~~~

#. Make sure that the advanced setting, ``kafka.auto_create_topics_enable`` is enabled. This can be set from
   the Overview tab for the Kafka service in the Aiven console.

> #. Enable Schema Registry from the Overview tab for the Kafka service

Reviewer 3: Do single sentence steps require punctuation?
```suggestion
#. Enable the Schema Registry from the Overview tab for the Kafka service
```

#. Create a topic named ``jdbc_sink``

> Download Kafka connect binaries:

Reviewer 3: ```suggestion
Download Kafka Connect binaries:
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#. Download the latest Kafka release from https://kafka.apache.org/quickstart

> #. Download the Aiven Kafka connect JDBC connector from https://github.com/aiven/jdbc-connector-for-apache-kafka/releases

Reviewer 3: ```suggestion
#. Download Aiven's JDBC Sink and Source Connectors for Apache Kafka from https://github.com/aiven/jdbc-connector-for-apache-kafka/releases
```
:D  We should never see "Aiven Kafka"

> #. Download the Avro Value Converter from https://www.confluent.io/hub/confluentinc/kafka-connect-avro-converter

Reviewer 3: ```suggestion
#. Download the Avro Converter from https://www.confluent.io/hub/confluentinc/kafka-connect-avro-converter
```

> Preparing Kafka connect software on a VM:

Reviewer 3: ```suggestion
Preparing Kafka Connect software on a VM:
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Log in to your VM.

::

   cd $HOME

   # extract kafka
   tar -xzf kafka_2.13-VERSION.tgz

   cd kafka_2.13-VERSION
   mkdir -p /share/kafka/plugins
   cd kafka/plugins
   # extract aiven connect jdbc
   unzip jdbc-connector-for-apache-kafka-VERSION.zip
   # extract confluent kafka connect avro converter
   unzip confluentinc-kafka-connect-avro-converter-VERSION.zip

|

> | Create a properties file, ``my-connect-distributed.properties`` , for Kafka connect.

Reviewer 3: ```suggestion
| Create a properties file, ``my-connect-distributed.properties``, for Kafka Connect.
```

   ssl.key.password=secret
   ssl.keystore.type=PKCS12

   consumer.ssl.endpoint.identification.algorithm=https
   consumer.request.timeout.ms=20000
   consumer.retry.backoff.ms=500
   consumer.security.protocol=SSL
   consumer.ssl.protocol=TLS
   consumer.ssl.truststore.location=/path/client.truststore.jks
   consumer.ssl.truststore.password=secret
   consumer.ssl.keystore.location=/path/client.keystore.p12
   consumer.ssl.keystore.password=secret
   consumer.ssl.key.password=secret
   consumer.ssl.keystore.type=PKCS12

   producer.ssl.endpoint.identification.algorithm=https
   producer.request.timeout.ms=20000
   producer.retry.backoff.ms=500
   producer.security.protocol=SSL
   producer.ssl.protocol=TLS
   producer.ssl.truststore.location=/path/client.truststore.jks
   producer.ssl.truststore.password=secret
   producer.ssl.keystore.location=/path/client.keystore.p12
   producer.ssl.keystore.password=secret
   producer.ssl.key.password=secret
   producer.ssl.keystore.type=PKCS12
   #

|
| **Import the Aiven project CA into the JVM's trust store.**

#. Download Aiven project CA - ca.pem.

Then transfer it to the VM. Execute the following steps on each VM

> participating in the connect cluster:

Reviewer 3: ```suggestion
participating in the Connect cluster:
```
